[MV-1195] Implement Pdist Backward

anhskrttt opened the pull request:
## Summary of Changes
- Add `PdistBackward` operation and kernel.
- Add driver test and gtest for `PdistBackward`.

### Additional Notes
- ROCm's `pdist` doesn't support for `dtype=[fp16, bfp16]` yet. This PR already includes support for those two types.
- MIOpen performance is better for contiguous inputs.
- For `input_dtype=fp16`, the result may experience underflow or overflow in the following cases: (1) when `input.dims[0]` is large, indicating a high number of points for distance calculation, or (2) when the value of `p` is large.

## Benchmark Results
### Average improvement over ROCm
| type     | bwd  |
|----------|------|
| float    | 1.61 |
| float16  | -  |
| bfloat16 | - |

### Detail Benchmark
Check the full benchmark result [here](https://morehio-my.sharepoint.com/:x:/r/personal/kyeonghwan_ryu_moreh_io/_layouts/15/Doc.aspx?sourcedoc=%7B498FA503-13F6-401A-9FFA-B3F736FBB0E2%7D&file=Dist%2C%20PDist%2C%20CDist.xlsx&action=default&mobileredirect=true).

hieule88 commented on driver/mloPdistHost.hpp:
```diff
@@ -0,0 +1,77 @@
+    auto M           = miopen::deref(inputDesc).GetLengths()[1];
+
+    // Fill dinputHost with zeros
+    for(size_t i = 0; i < input_numel; i++)
+    {
+        dinputHost[i] = 0;
+    }
+
+    for(int i = 0; i < N; ++i)
+    {
+        for(int j = i + 1; j < N; ++j)
```
pard_for for better perf

hieule88 replied:
use uint64_t i, j

anhskrttt replied:
What's the reason to use `uint64_t`? If it was about the semantic problem, then I think using `size_t` seem to be more proper?

anhskrttt replied:
I don't think we can use par_ford() here, because I need to accumulate the results later. If we use parallel computing, the accumulation would require atomic operations, which would complicate things:

```
dinputHost[i * M + m] += res;
dinputHost[j * M + m] -= res;
``` 

An alternative solution that would allow the use of parallel computing, is by creating a temporary intermediate variable to store the partial sums (Calculate this partial sum can use parallel computatuon). And finally,  do a sumReduction step to compute the final result—similar to the approach used in the kernel code. However, I don’t think this is necessary since the purpose of this function is to verify correctness rather than focus on performance.

If you notice any opportunities for parallelization in this code, please point them out specifically. At the moment, I don’t see how it could be computed in parallel. Thanks.

hieule88 replied:
N is size_t (uint64_t) while i,j is still int which leads to failed CI/CD test. use size_t is fine

hieule88 commented on driver/mloPdistHost.hpp:
```diff
@@ -0,0 +1,77 @@
+            double grad_k   = static_cast<double>(doutput[k]);
+            double output_k = static_cast<double>(output[k]);
+
+            for(int m = 0; m < M; ++m)
```
this may lead to CI/CD failed because M is size_t while m is int (comparation between different type)

anhskrttt replied:
I'll change dtype of both `m` and `M` into `size_t` to avoid unexpected behavior of comparing between different types.

hieule88 replied:
did you change it ?

hieule88 commented on driver/mloPdistHost.hpp:
```diff
@@ -0,0 +1,77 @@
+
+            for(int m = 0; m < M; ++m)
+            {
+                double input_first  = static_cast<double>(input[i * M + m]);
+                double input_second = static_cast<double>(input[j * M + m]);
+                double diff         = input_first - input_second;
+
+                Tcheck res =
+                    static_cast<Tcheck>(miopen::solver::pdist::backward(diff, grad_k, output_k, p));
+
+                dinputHost[i * M + m] += res;
+                dinputHost[j * M + m] -= res;
```
use tensor_view instead: tv.get_tensor_view_idx({i, m})

anhskrttt replied:
This implementation only covers the contiguous version, so I didn't need to use tensor_view.

The reason I initially focused only on the non-contiguous version was due to the benchmark results. However, after a careful review of my PRs, Hieu-nim suggested adding more cases for non-contiguous inputs to determine if the non-contiguous version still offers performance improvements. This will be addressed in a separate ticket/commit. For now, I'll mark this as resolved since this PR only handles the contiguous input version.

hieule88 replied:
This is for future improvement if needed. non-con cpu code can both apply to check non-con and con kernel. And please do not resolve review yourself

anhskrttt replied:
TODO: Change verification cpu code from contiguous only into non-contiguous implementation, because non-cont cpu code can be applied and used for testing for both cont and non-cont kernel.

hieule88 commented on driver/mloPdistHost.hpp:
```diff
@@ -0,0 +1,77 @@
+    {
+        for(int j = i + 1; j < N; ++j)
+        {
+            long k          = j + N * i - i * (i + 1) / 2 - i - 1;
```
can k be negative? if not, use uint64_t

hieule88 replied:
if can be, check for k before use it as index

anhskrttt replied:
k cannot be negative, I'll fix its type

hieule88 commented on driver/pdist_driver.hpp:
```diff
@@ -0,0 +1,363 @@
+ *******************************************************************************/
+#pragma once
+
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <memory>
+#include <vector>
```
are those lib necessary ?

hieule88 replied:
follow this order (AMD's convention):
`#include "InputFlags.hpp"
#include "driver.hpp"
#include "mloMarginRakningLossHost.hpp"
#include "random.hpp"
#include "tensor_driver.hpp"
#include "timer.hpp"

#include <../test/tensor_holder.hpp>
#include <../test/verify.hpp>

#include <miopen/env.hpp>
#include <miopen/handle.hpp>
#include <miopen/miopen.h>
#include <miopen/tensor.hpp>
#include <vector>`

anhskrttt replied:
those libs are automatically added by one of the vscode's extensions. I'll remove them

anhskrttt replied:
Could you help me locate AMD's official documentation or guidelines regarding library include order? I've been unable to find a definitive reference, and the current include structure in our files seems somewhat inconsistent. Thanks.

hieule88 replied:
this AMD's convention from long time ago. Please use TrustMeBro for reference

hieule88 commented on driver/pdist_driver.hpp:
```diff
@@ -0,0 +1,363 @@
+template <typename Tgpu, typename Tref>
+int PdistDriver<Tgpu, Tref>::AddCmdLineArgs()
+{
+    inflags.AddInputFlag("forw", 'F', "0", "Run only Pdist Backward (Default=0)", "int");
```
I find you make no additional setup for this -F 0 for only backward. In default set up: -F 0 is for running both Forward and Backward. -F 1 is for Forward and -F 2 is for Backward. Here is the result of your op when using -F 0:
 
<img width="557" alt="image" src="https://github.com/user-attachments/assets/bbe8b985-5780-4073-8b85-6d783d7b5128">

anhskrttt replied:
i misunderstood that `-F 0` is for backward. I fixed this to `-F 2`, which indicates only run backward

anhskrttt commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+{
+    pdist_backward_contiguous<INPUT_TYPE>(
+        input, output, grad, input_grad, p, n2, n2_squared_minus_1, N, NO, M);
+}
```
add new line at the end

hieule88 commented on driver/mloPdistHost.hpp:
```diff
@@ -0,0 +1,77 @@
+    for(size_t i = 0; i < input_numel; i++)
+    {
+        dinputHost[i] = 0;
+    }
```
std::fill(dinputHost, dinputHost + input_numel, 0.0);

anhskrttt replied:
resolved

hieule88 commented on driver/pdist_driver.hpp:
```diff
@@ -0,0 +1,363 @@
+                                           std::numeric_limits<Tgpu>::max());
+    }
+
+    std::fill(doutput.begin(), doutput.end(), 1.0);
```
why this param is not gen as above params

anhskrttt replied:
`1.0` is simply a random value. I noticed that other operations also use this value, so I followed the same approach. Would it be better to use gen_value as the `input` tensor above instead? If there's no significant difference, let's keep it as it is for now.

hieule88 commented on src/include/miopen/pdist.hpp:
```diff
@@ -0,0 +1,55 @@
+
+struct Handle;
+struct TensorDescriptor;
+
```
add namespace pdist

anhskrttt replied:
resolved

hieule88 commented on src/include/miopen/pdist/problem_description.hpp:
```diff
@@ -0,0 +1,149 @@
+ *******************************************************************************/
+#pragma once
+
+#include <cstddef>
+
+#include <miopen/errors.hpp>
+#include <miopen/miopen.h>
+#include <miopen/problem_description_base.hpp>
+#include <miopen/activ.hpp>
+#include <miopen/tensor.hpp>
```
remove unuse libs and re-order as alphabet

anhskrttt replied:
resolved

hieule88 commented on src/pdist/problem_description.cpp:
```diff
@@ -0,0 +1,51 @@
+    ss << "pdist_backward";
+    ss << "dtype" << input_dtype;
+    ss << "dims" << input_dims[0] << "x" << input_dims[1];
+    ss << "p" << p;
```
this is redundant because its not appear in IsApplicable

anhskrttt replied:
I'm not fully familiar with the `MakeNetworkConfig()`, so just to make sure, I want to confirm that the redundancy is specifically related to the include of `p`, and that removing line 44 would resolve the issue completely, is this correct?

anhskrttt replied:
resolved: remove `p` since its not included in `IsApplicable()` or `GetSolution()`. Therefore, there's no need to include `p` in `NetworkConfigs`

hieule88 commented on src/include/miopen/pdist/problem_description.hpp:
```diff
@@ -0,0 +1,149 @@
+        if(inputDesc.GetNumDims() != 2)
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "PdistBackward: Input tensor must be 2D.");
+            return false;
```
redundant

hieule88 replied:
you missed this return false?

anhskrttt replied:
sorry for the carelessness. I removed this.

hieule88 commented on src/include/miopen/pdist/problem_description.hpp:
```diff
@@ -0,0 +1,149 @@
+                             std::to_string(expected_size) +
+                             ". Got: " + std::to_string(actual_num_dims) + "D tensor of size " +
+                             std::to_string(actual_size) + ".");
+            return false;
```
redundant

anhskrttt replied:
solved

hieule88 commented on src/include/miopen/pdist/problem_description.hpp:
```diff
@@ -0,0 +1,149 @@
+             dinputDesc.IsContiguous()))
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "PdistBackward: Uncontiguous tensors not supported.");
+            return false;
```
redundant

anhskrttt replied:
solved

hieule88 commented on src/include/miopen/pdist/problem_description.hpp:
```diff
@@ -0,0 +1,149 @@
+             dinputDesc.IsPacked()))
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "PdistBackward: Unpacked tensors not supported.");
+            return false;
```
redundant

anhskrttt replied:
solved

hieule88 commented on src/solver/pdist/backward_pdist.cpp:
```diff
@@ -0,0 +1,284 @@
+{
+    std::ignore = context;
+
+    if(!problem.IsAllContiguous())
+    {
+        return false;
+    }
+
+    if(!problem.IsAllPacked())
+    {
+        return false;
+    }
```
this should be in problem description constructor. check https://moreh.atlassian.net/wiki/spaces/MV/pages/903512367/MV-GPU-MIOpen+NetworkConfig+example for more information

anhskrttt replied:
UPDATED: I consulted Hieu-nim again, it turned out that I had missed adding `IsAllContiguous()` to `NetworkConfigs`, leading to incorrect configurations later. Hieu-nim suggested moving this to the constructor for better semantic clarity.

--
I'm still a little confused about the `NetworkConfigs` so I want to make clear about this again. Please confirm that my understanding is correct: Given that I've only implemented the contiguous version of `Pdist`, the `IsAllContiguous()` check should be moved into `ProblemDescription`. This would ensure that non-contiguous cases fail at the problem construction stage, as they aren't valid problems under the current implementation.

hieule88 commented on src/include/miopen/pdist/problem_description.hpp:
```diff
@@ -0,0 +1,149 @@
+        return true;
+    }
+
+    bool IsAllPacked() const
+    {
+        if(!(inputDesc.IsPacked() && outputDesc.IsPacked() && doutputDesc.IsPacked() &&
+             dinputDesc.IsPacked()))
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "PdistBackward: Unpacked tensors not supported.");
+            return false;
```
if its contiguous, its obviously packed

anhskrttt replied:
solved: removed this condition

hieule88 commented on src/include/miopen/pdist/solvers.hpp:
```diff
@@ -0,0 +1,60 @@
+ *******************************************************************************/
+#pragma once
+
+#include "miopen/conv_solution.hpp"
```
#include <miopen

anhskrttt replied:
solved

hieule88 commented on src/include/miopen/pdist/utils.hpp:
```diff
@@ -0,0 +1,69 @@
+#include <math.h>
+
+namespace miopen {
+namespace solver {
```
why those code should be in "solver" namespace?

anhskrttt replied:
Using the `solver` namespace was overly specific for this code, so I'll remove it. These utilities only need to be under the `miopen::pdist` namespace

hieule88 commented on src/include/miopen/pdist/utils.hpp:
```diff
@@ -0,0 +1,69 @@
+namespace solver {
+namespace pdist {
+
+inline bool is_approx_equal(double a, double b, double tolerance = 1e-6)
+{
+    return (fabs(a - b) < tolerance);
+}
+
+inline double sign_(double val) { return (0 < val) - (val < 0); }
+
+inline double backward(const double diff, const double grad, const double dist, const double p)
```
you can use Template type T for general

anhskrttt replied:
solved: changed those functions into template <type T>

hieule88 commented on src/include/miopen/solver_id.hpp:
```diff
@@ -61,7 +61,8 @@ enum class Primitive
     Item,
     RoPE,
-    ReLU
+    ReLU,
+    Pdist
```
update code with branch develop-moreh

anhskrttt replied:
solved

This happened because I initially checked out the wrong branch (`develop`) when starting to write the new operation—it should have been `develop-moreh`. I resolved the issue by merging the changes with `develop-moreh` and resolving all conflicts.

hieule88 commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+
+    FLOAT_ACCUM res = backward(diff, grad_k, output_k, p_);
+
+    input_grad[ib * N * M + i * M + m] = CVT_ACCUM2FLOAT(res);  // dist(i,j)
+    input_grad[jb * N * M + j * M + m] = CVT_ACCUM2FLOAT(-res); // dist(j,i)
```
Why you use this type of indexing why CL code use tensor_view? 
```
TV_3D_AT(input_grad, ib, i, m) = res;
TV_3D_AT(input_grad, jb, j, m) = -res; 
```

hieule88 commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+    // pair(row_i, row_j) corresponding to output[k]
+    // e.g. pair(point_A, point_B)
+    long i = n2 - sqrt(n2_squared_minus_1 - 2 * k);
+    long j = k - N * i + i * (i + 1) / 2 + i + 1;
+
+    // Pair of gradients corresponding to pair(i,j)
+    // e.g. pair(input_grad(A,B), input_grad(B,A))
+    // those 2 elements in pair have opposite signs
+    // just similar to dist(A,B) vs. dist(B,A)
+    long ib = j - i - 1;
+    long jb = N - 2 - i;
+
+    FLOAT_ACCUM grad_k   = CVT_FLOAT2ACCUM(grad[k]);
+    FLOAT_ACCUM output_k = CVT_FLOAT2ACCUM(output[k]);
+
+    FLOAT_ACCUM diff = CVT_FLOAT2ACCUM(input[i * M + m]) - CVT_FLOAT2ACCUM(input[j * M + m]);
+
+    FLOAT_ACCUM p_ = static_cast<FLOAT_ACCUM>(p);
+
+    FLOAT_ACCUM res = backward(diff, grad_k, output_k, p_);
+
+    input_grad[ib * N * M + i * M + m] = CVT_ACCUM2FLOAT(res);  // dist(i,j)
+    input_grad[jb * N * M + j * M + m] = CVT_ACCUM2FLOAT(-res); // dist(j,i)
+}
+
+extern "C" __global__ void
+PdistBackwardContiguous(const INPUT_TYPE* __restrict__ input,
+                        const INPUT_TYPE* __restrict__ output,
+                        const INPUT_TYPE* __restrict__ grad, // output_grad
+                        INPUT_TYPE* __restrict__ input_grad,
+                        double p,
+                        double n2,
+                        double n2_squared_minus_1,
+                        long N,
+                        long NO,
+                        long M)
+{
+    pdist_backward_contiguous<INPUT_TYPE>(
+        input, output, grad, input_grad, p, n2, n2_squared_minus_1, N, NO, M);
+}
```
I find CL code do not use this contiguous func and this make no difference with the above function because you use contiguous indexing input_grad[ib * N * M + i * M + m] in the above function

anhskrttt replied:
TODO: Fix non-cont implementation

hieule88 commented on src/pdist.cpp:
```diff
@@ -0,0 +1,103 @@
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include <miopen/execution_context.hpp>
+#include <miopen/miopen.h>
+#include <miopen/names.hpp>
+#include <miopen/datatype.hpp>
```
may be redundant

anhskrttt replied:
i removed those redundant libs

hieule88 commented on src/pdist.cpp:
```diff
@@ -0,0 +1,103 @@
+#include <miopen/pdist/problem_description.hpp>
+
+namespace miopen {
+
```
add namespace pdist

anhskrttt replied:
added namespace pdist

hieule88 commented on src/pdist_api.cpp:
```diff
@@ -0,0 +1,135 @@
+#include <miopen/tensor_ops.hpp>
+#include <miopen/pdist.hpp>
+
+static void LogCmdPdistBackward(const miopenTensorDescriptor_t inputDesc, const double p)
```
make it inline function

anhskrttt replied:
can you point out the reason why this function should be make as inline function? I see other `LogCmd` functions are just  normal `void` function instead of `inline` function.

p/s: I changed this into inline function. But pls still explain the reason why for me.

hieule88 commented on src/pdist_api.cpp:
```diff
@@ -0,0 +1,135 @@
+        }
+
+        ss << "-dims ( ";
+        for(unsigned long i : miopen::deref(inputDesc).GetLengths())
```
uint64_t

anhskrttt replied:
same as I discuss above, I think using `size_t` seems more proper? Please share your thoughts on this?

hieule88 replied:
both uint64_t and size_t are fine but dont use unsigned long (which does not guarantee 64 bit or big enough to contain the size of the biggest object your system can handle)

hieule88 commented on src/pdist_api.cpp:
```diff
@@ -0,0 +1,135 @@
+
+extern "C" miopenStatus_t miopenPdistBackward(miopenHandle_t handle,
+                                              void* workspace,
+                                              size_t workspaceSizeInBytes,
```
add const

anhskrttt replied:
solved

hieule88 commented on src/solver.cpp:
```diff
@@ -34,6 +34,7 @@
 #include <miopen/groupnorm/solvers.hpp>
 #include <miopen/getitem/solvers.hpp>
 #include <miopen/layernorm/solvers.hpp>
+#include "miopen/pdist/solvers.hpp"
```
#include <miopen

anhskrttt replied:
solved

hieule88 commented on src/solver/pdist/backward_pdist.cpp:
```diff
@@ -0,0 +1,284 @@
+#include <vector>
+
+#include <miopen/miopen.h>
+#include "miopen/conv_solution.hpp"
```
#include <miopen

anhskrttt replied:
solved

hieule88 commented on src/solver/pdist/backward_pdist.cpp:
```diff
@@ -0,0 +1,284 @@
+#include <miopen/kernel_info.hpp>
+#include <miopen/mlo_internal.hpp>
+#include <miopen/target_properties.hpp>
+#include "miopen/buffer_info.hpp"
+#include "miopen/errors.hpp"
+#include "miopen/execution_context.hpp"
+#include "miopen/invoke_params.hpp"
+#include "miopen/tensor.hpp"
+#include "miopen/tensor_view_utils.hpp"
```
same

anhskrttt replied:
solved

hieule88 commented on src/solver/pdist/backward_pdist.cpp:
```diff
@@ -0,0 +1,284 @@
+#include <miopen/pdist.hpp>
+#include <miopen/pdist/solvers.hpp>
+#include <miopen/pdist/invoke_params.hpp>
+#include "miopen/pdist/problem_description.hpp"
```
same

anhskrttt replied:
solved

hieule88 commented on src/solver/pdist/backward_pdist.cpp:
```diff
@@ -0,0 +1,284 @@
+{
+    std::ignore = context;
+
+    if(!problem.IsAllContiguous())
+    {
+        return false;
+    }
```
I found no kernel implemented for non-contiguous case, please add and re-benchmark the performance. btw, I see you define problem.IsAllContiguou() as not supported, this is not a clear meaning. If you did benchmark the right implementation of non-contiguous kernel, this IsAllContiguou() should not THROW, just return false and should be in IsOverRocm function

anhskrttt replied:
Currently, I've only implement kernel for cont inputs. Hence, IsAllContiguous should THROW.

TODO: Add non-cont solution and rebenchmark

hieule88 commented on src/solver/pdist/backward_pdist.cpp:
```diff
@@ -0,0 +1,284 @@
+
+    auto dtype = problem.GetInputDesc().GetType();
+
+    auto input_dtype  = miopen::GetDataType(problem.GetdInputDesc().GetType());
+    auto output_dtype = miopen::GetDataType(problem.GetdOutputDesc().GetType());
+    auto dinput_dtype = miopen::GetDataType(problem.GetdInputDesc().GetType());
```
you have IsSameType so this is redundant

anhskrttt replied:
I understand that these types are all the same. However, I intentionally query each one separately to enhance code readability and self-documentation. For example, I would like to use `{"OUTPUT_TYPE", dinput_dtype == "bfloat16" ? "ushort" : dinput_dtype}` to explicitly shows that the output type is `dinput_dtype`, which is clearer than using `{"OUTPUT_TYPE", input_dtype == "bfloat16" ? "ushort" : input_dtype}` (even when I know they are all the same). I think it would be easier for future readers to understand the code.

Pls take your time to review my fix for this and consider if it's ok now. If it's not, pls let me know

hieule88 replied:
you can use variable name: dtype then use for all variable, still ensure readability

anhskrttt replied:
solved:
```
auto dtype = problem.GetInputDesc().GetType();
auto dtype_str = miopen::GetDataType(dtype);
```

hieule88 commented on src/solver/pdist/backward_pdist.cpp:
```diff
@@ -0,0 +1,284 @@
+            {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+            {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+            {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+            {"INPUT_TYPE", input_dtype == "bfloat16" ? "ushort" : input_dtype},
+            {"OUTPUT_TYPE", input_dtype == "bfloat16" ? "ushort" : output_dtype},
```
typo?

anhskrttt replied:
I fixed this.

Pls note here that while I am required to specify both input and output types because I'm using the external pre-defined kernel `CalculationFwdContiguous`, for the `pdist` operation specifically, these types are identical. Another thing to note is that, for this kernel, my output is `dinput`, so I'll pass `dinput_type` to `OUTPUT_TYPE`. The names might be a little confused.

hieule88 replied:
use dtype for both INPUT_TYPE and OUTPUT_TYPE

anhskrttt replied:
solved

hieule88 reviewed (changes requested):
[Partial Review] I found no kernel rightly implemented for non-contiguous case, please add and re-benchmark the performance. Additionally, why you dont implement Forward case, I found pdist kernel (forward case) in CL code and before all of the modification you will make, please change base branch to develop-moreh for the most update code

anhskrttt commented:
> why you dont implement Forward case, I found pdist kernel (forward case) in CL code

Initially, I didn't implement Forward cases because our benchmarking of ROCm versus modnn showed no performance improvements in Forward operations. The PR description noted that performance gains were only observed with backward operations and contiguous input tensors.

However, upon further review, I recognize the need for more comprehensive testing. I plan to expand our benchmarking to include both contiguous and non-contiguous cases to definitively verify whether the performance improvements are indeed limited to contiguous cases only. This additional testing will help us make a more informed decision about implementation scope.

anhskrttt commented:
Ticket MV-1195 will focus specifically on implementing the contiguous version of PdistBackward. The non-contiguous implementation will be addressed separately in a new ticket to maintain clear scope and trackability.
cc @long10024070

hieule88 commented on test/cpu_pdist.hpp:
```diff
@@ -0,0 +1,70 @@
+    auto N = input.desc.GetLengths()[0];
+    auto M = input.desc.GetLengths()[1];
+
+    for(int i = 0; i < N; ++i)
+    {
+        for(int j = i + 1; j < N; ++j)
+        {
+            long k = j + N * i - i * (i + 1) / 2 - i - 1;
+
+            double grad_k   = static_cast<double>(output_grad[k]);
+            double output_k = static_cast<double>(output[k]);
+
+            for(int m = 0; m < M; ++m)
```
same as mlo

anhskrttt replied:
TODO: Change verification cpu code from contiguous only into non-contiguous implementation, because non-cont cpu code can be applied and used for testing for both cont and non-cont kernel.

hieule88 commented on test/gtest/pdist.cpp:
```diff
@@ -0,0 +1,62 @@
+    Verify();
+};
+
+INSTANTIATE_TEST_SUITE_P(Smoke, GPU_Pdist_bwd_FP32, testing::ValuesIn(PdistTestConfigs()));
+INSTANTIATE_TEST_SUITE_P(Smoke, GPU_Pdist_bwd_FP16, testing::ValuesIn(PdistFp16TestConfigs()));
+INSTANTIATE_TEST_SUITE_P(Smoke, GPU_Pdist_bwd_BFP16, testing::ValuesIn(PdistTestConfigs()));
+
+INSTANTIATE_TEST_SUITE_P(Full, GPU_Pdist_bwd_FP32, testing::ValuesIn(PdistTestConfigs()));
+INSTANTIATE_TEST_SUITE_P(Full, GPU_Pdist_bwd_FP16, testing::ValuesIn(PdistFp16TestConfigs()));
+INSTANTIATE_TEST_SUITE_P(Full, GPU_Pdist_bwd_BFP16, testing::ValuesIn(PdistTestConfigs()));
+
+INSTANTIATE_TEST_SUITE_P(Perf, GPU_Pdist_bwd_FP32, testing::ValuesIn(PdistTestConfigs()));
+INSTANTIATE_TEST_SUITE_P(Perf, GPU_Pdist_bwd_FP16, testing::ValuesIn(PdistFp16TestConfigs()));
+INSTANTIATE_TEST_SUITE_P(Perf, GPU_Pdist_bwd_BFP16, testing::ValuesIn(PdistTestConfigs()));
```
```suggestion
INSTANTIATE_TEST_SUITE_P(Full, GPU_Pdist_bwd_FP32, testing::ValuesIn(PdistTestConfigs()));
INSTANTIATE_TEST_SUITE_P(Full, GPU_Pdist_bwd_FP16, testing::ValuesIn(PdistFp16TestConfigs()));
INSTANTIATE_TEST_SUITE_P(Full, GPU_Pdist_bwd_BFP16, testing::ValuesIn(PdistTestConfigs()));
```

anhskrttt replied:
I fixed this.

My additional question: Is there any references that say we only need to do test_type of `Full`. I included all types of test because I refered to the op `prelu`.

hieule88 replied:
you have only one test set, so it is of TEST_TYPE = Full. Check https://github.com/ROCm/MIOpen/wiki/GTest-development for more in formation

hieule88 commented on test/gtest/pdist.hpp:
```diff
@@ -0,0 +1,228 @@
+        
+    };
+    // clang-format on
+}
```
why you separate test cases to 2 test ?

anhskrttt replied:
Large `p` causes fp16 cases to be failed due to overflow/underflow sometimes, so I remove large p in Fp16TestConfig (e.g. p=5)

hieule88 commented on test/gtest/pdist.hpp:
```diff
@@ -0,0 +1,228 @@
+        auto N               = input_dims[0];
+        auto output_dim_size = N * (N - 1) / 2;
+
+        if(output_dim_size == 0)
+        {
+            GTEST_SKIP();
+        }
```
if this is condition of the ops, add it to problem description and remove this

anhskrttt replied:
MIOpen currently does not support handling empty tensors. However, my operation includes scenarios where `input.shape[0] == 0` or `input.shape[0] == 1`, in which case the output should be an empty tensor. In these situations, I'm unable to create an output tensor necessary to run `miopen::pdist::PdistBackward`. This limitation prevents me from even reaching the `ProblemDescription` stage, as creating a `TensorDescription` for an output with `dim_size = 0` isn't possible. I have kept the relevant lines of code in place as they might support empty tensors in the future and to serve as a reminder to handle cases where `input.shape[0] == 1`. Do you think I should simply remove those cases if they are not supported yet?

Notice that op `masked_select` has similar cases to handle. An example of op `masked_select` to return an empty tensor:
```
x = torch.tensor([1, 2, 3])
mask = torch.tensor([False, False, False])
result = torch.masked_select(x, mask)  # tensor([])
```
But this op hasnt been implemented yet.

My current solution is just simply throw an exception everytime we occur to need to handle empty tensors (input/output). Do you have any addition suggestions on this?

hieule88 replied:
actually, no previous op handled empty tensors. Let AMD decide

hieule88 commented on test/gtest/pdist.hpp:
```diff
@@ -0,0 +1,228 @@
+                                       dinput_dev.get(),
+                                       p);
+
+        EXPECT_EQ(status, miopenStatusSuccess);
```
```suggestion
        ASSERT_EQ(status, miopenStatusSuccess);
```

anhskrttt replied:
solved

hieule88 commented on test/gtest/pdist.hpp:
```diff
@@ -0,0 +1,228 @@
+    {
+        double threshold  = std::numeric_limits<T>::epsilon() * 10;
+        auto dinput_error = miopen::rms_range(ref_dinput, dinput);
+
```
ASSERT_EQ(miopen::range_distance(ref_dinput), miopen::range_distance(dinput));

anhskrttt replied:
`range_distance()` is already covered in `rms_range()` so I think check it again is duplicated. Please refer to the `rms_range()` implementation:

```
template <class R1, class R2>
double rms_range(R1&& r1, R2&& r2)
{
    std::size_t n = range_distance(r1);
    if(n == range_distance(r2))
    {
        if(n == 0)
            return 0;
        double square_difference = range_product(r1, r2, 0.0, sum_fn{}, square_diff);
        double mag1 = static_cast<double>(*std::max_element(r1.begin(), r1.end(), compare_mag));
        double mag2 = static_cast<double>(*std::max_element(r2.begin(), r2.end(), compare_mag));
        double mag =
            std::max({std::fabs(mag1), std::fabs(mag2), std::numeric_limits<double>::min()});
        return std::sqrt(square_difference) / (std::sqrt(n) * mag);
    }
    else
        return double(std::numeric_limits<range_value<R1>>::max());
}
```

hieule88 replied:
rms_range return max limit when range_distance() failed while ASSERT_EQ throw the range_distance() of each output

anhskrttt replied:
If `rms_range` returns the maximum limit, it will certainly fail the `EXPECT_LT(dinput_error, threshold)` check if the range check is failed. I guess the reason `range_distance` was included inside `rms_range` is to remove the need for an additional `range_distance` check outside of this function, as `rms_range` already covers it. Using `ASSERT_EQ` is fine, but if it's not a strict convention pointed out by AMD side to always check `range_distance` even when `rms_range` is utilized, then the external check seems redundant. (If there's any reference for this requirement, pls let me know)

However, I also notice that, many other ops also do both `ASSERT_EQ` range distance in addition to `rms_range`. So I'm not sure if I need it or not..? (imo, i still think it's redundant)

cc @long10024070 @littlecutebird @cognaiger9

hieule88 replied:
as I said, ASSERT_EQ throw the range_distance() (different between distance from last value and begin value - maybe number of elements, view,...) of 2 outputs while rms_range() show the different between values of 2 outputs. Exp: if somehow, you forgot init ref_dinput=tensor<T>{...} so you will get range_distance(ref_dinput) =0, ASSERT_EQ will show you that while rms_range() just show you max litmit value (nonsense in this case). They show the different side of bugs, so each one have their own separated meaning.

anhskrttt replied:
My suggestion is just simple that `rms_range()` alone would be sufficient to trigger the exception. I understand that keeping `ASSERT_EQ(range_distance())` provides more detailed diagnostics, because it helps identify whether the failure occurred due to a range distance mismatch or an `rms_range` issue by showing the specific value differences. However, pls note that, the reason I raised this, is because running both `ASSERT_EQ(range_distance())` and `rms_range()` creates redundant validation logic and therefore, this presents an opportunity to improve the `rms_range()` function in the future. But since it's not a critical issue, let's just follow your suggestion.

cc @cognaiger9: I previously suggested removing the `range_distance` check in your PR. Please refer to this discussion and re-add it the for better error clarity if needed. Sorry for the inconvenience.

hieule88 reviewed (commented):
Full review

anhskrttt commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+}
+
+template <typename DTYPE>
+__device__ void pdist_backward(const DTYPE* __restrict__ input,
```
check non-cont solution again

anhskrttt replied:
updated: implemented non-cont solution. Consult Hieu-nim to decide if keep non-cont or rm. but basically, this comment is solved. 

solved: ~non-cont solution doesnt improve perf so I'll remove non-cont kernel implementation~

hieule88 commented on test/gtest/pdist.hpp:
```diff
@@ -0,0 +1,228 @@
+{
+    return {
+        // TODO: How to test when expected output=tensor([])?
+        // PdistTestCase({1, 1}, 0.0),
+        // PdistTestCase({1, 1}, 1.0),
+        // PdistTestCase({1, 1}, 2.0),
+        // PdistTestCase({1, 1}, 5.0),
+        // PdistTestCase({1, 1}, HUGE_VAL),
```
remove comments

anhskrttt replied:
Those tests are for cases when the output is empty tensor: `tensor([])`. I'm fixing my op for those cases. I'll update this asap

anhskrttt replied:
pls check my discussion [here](https://github.com/moreh-dev/MIOpen/pull/62#discussion_r1855749335)

anhskrttt replied:
solved: throw when input/output is empty tensor

anhskrttt commented:
I implemented the `PdistBwd` for non-cont input. Even those for small tests, I still think the performance seems not to be very promising so I'll remove non-cont implementation in this op. Check non-cont benchmark result [here](https://1drv.ms/x/c/f55f630f37e718b8/EXv4O1zOO31IgQmkQyMw5RwBTR-zbeHKMuGgmKzdozibAQ?e=ZOshun&nav=MTVfezZERURBNzc3LTU2OUMtNEI2NS1CRkU5LTM2QTMyODEzRkMzNX0).
![image](https://github.com/user-attachments/assets/f24bd41e-f9af-4476-8f25-11171198a8df)

cc. @hieule88

anhskrttt commented:
UPDATED: Non-cont cases seems to be as good as cont cases. Pls check my update in [this comment](https://github.com/moreh-dev/MIOpen/pull/62#issuecomment-2503719826).

@hieule88  I have added non-cont implementation to this PR (to re-benchmark perf). ~I think non-cont doesnt offer promising performance enhancement. Should I remove the non-cont implement in this PR, or just keep it and move `IsAllContiguous()` condition into `IsApplicable` instead of `ProblemDescription`?~

hieule88 commented on src/include/miopen/pdist/problem_description.hpp:
```diff
@@ -0,0 +1,138 @@
+        if(!(inputDesc.IsContiguous() && outputDesc.IsContiguous() && doutputDesc.IsContiguous() &&
+             dinputDesc.IsContiguous()))
+        {
+            // MIOPEN_THROW(miopenStatusBadParm, "PdistBackward: All tensors must be contiguous.");
```
remove comments

anhskrttt replied:
solved

hieule88 commented on src/solver.cpp:
```diff
@@ -686,7 +687,6 @@ inline SolverRegistrar::SolverRegistrar(IdRegistryData& registry)
 
     Register(registry, ++id, Primitive::Activation, glu::GLUForward{}.SolverDbId());
     Register(registry, ++id, Primitive::Activation, glu::GLUBackward{}.SolverDbId());
-
```
unintended change

anhskrttt replied:
solved

hieule88 commented on src/solver/pdist/backward_pdist.cpp:
```diff
@@ -0,0 +1,441 @@
+{
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    // auto is_contiguous = problem.GetInputDesc().IsContiguous();
```
remove comment

anhskrttt replied:
solved

hieule88 commented on test/gtest/pdist.cpp:
```diff
@@ -0,0 +1,54 @@
+using GPU_Pdist_bwd_FP16  = PdistTestBackward<float16>;
+using GPU_Pdist_bwd_BFP16 = PdistTestBackward<bfloat16>;
+
+TEST_P(GPU_Pdist_bwd_FP32, PdistTestBackward)
```
```suggestion
TEST_P(GPU_Pdist_bwd_FP32, Test)
```

anhskrttt replied:
solved, however, i would like to ask the reason why change it to `Test`. This 2nd param of `TEST_P()` is just simply a TestName, I thought we can choose any test name we want. Is there any refs specify we have to use "Test" for 2nd of `TEST_P()`?

hieule88 commented on test/gtest/pdist.cpp:
```diff
@@ -0,0 +1,54 @@
+    Verify();
+};
+
+TEST_P(GPU_Pdist_bwd_FP16, PdistTestBackward)
+{
+    RunTest();
+    Verify();
+};
+
+TEST_P(GPU_Pdist_bwd_BFP16, PdistTestBackward)
```
same

anhskrttt replied:
solved

hieule88 commented on test/gtest/pdist.hpp:
```diff
@@ -0,0 +1,240 @@
+    };
+}
+
+// Remove big tests with reduction from FP16 test because the result will be overflow/ underflow
```
This may because of the too big value from gen_descreet_uniform_sign. try 
```
auto gen_value = [](auto...) { return prng::gen_descreet_uniform_sign<T>(1e-2, 1); };
        auto output_gen_value = [](auto...) {
            return prng::gen_descreet_uniform_sign<T>(1e-2, 1);
        };
```
Then remove PdistFp16TestConfigs because of the duplicated

anhskrttt replied:
Solved: I got it. I fixed as suggested and all test cases passed. I'll remove `PdistFp16TestConfigs`.

hieule88 commented on test/cpu_pdist.hpp:
```diff
@@ -0,0 +1,79 @@
+    auto N = input.desc.GetLengths()[0];
+    auto M = input.desc.GetLengths()[1];
+
+    for(int i = 0; i < N; ++i)
+    {
+        for(int j = i + 1; j < N; ++j)
+        {
```
int to size_t

anhskrttt replied:
solved

hieule88 commented on test/cpu_pdist.hpp:
```diff
@@ -0,0 +1,79 @@
+    {
+        for(int j = i + 1; j < N; ++j)
+        {
+            long k = j + N * i - i * (i + 1) / 2 - i - 1;
```
long to size_t (because you said k can be negative)

anhskrttt replied:
*can't

anhskrttt replied:
solved

hieule88 commented on test/cpu_pdist.hpp:
```diff
@@ -0,0 +1,79 @@
+                static_cast<double>(output_grad[output_grad_tv.get_tensor_view_idx({k})]);
+            double output_k = static_cast<double>(output[output_tv.get_tensor_view_idx({k})]);
+
+            for(int m = 0; m < M; ++m)
```
int to size_t

anhskrttt replied:
solved

hieule88 commented on src/solver/pdist/backward_pdist.cpp:
```diff
@@ -0,0 +1,441 @@
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include "miopen/tensor_view_utils.hpp"
```
include <miopen

anhskrttt replied:
solved

hieule88 commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+    input_grad[jb * N * M + j * M + m] = CVT_ACCUM2FLOAT(-res); // dist(j,i)
+}
+
+extern "C" __global__ void PdistBackward(const INPUT_TYPE* __restrict__ input,
+                                         const INPUT_TYPE* __restrict__ output,
+                                         const INPUT_TYPE* __restrict__ grad, // output_grad
+                                         INPUT_TYPE* __restrict__ input_grad,
```
change INPUT_TYPE to D_TYPE for readability

anhskrttt replied:
solved

hieule88 commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+}
+
+extern "C" __global__ void
+PdistBackwardContiguous(const INPUT_TYPE* __restrict__ input,
+                        const INPUT_TYPE* __restrict__ output,
+                        const INPUT_TYPE* __restrict__ grad, // output_grad
+                        INPUT_TYPE* __restrict__ input_grad,
```
same

anhskrttt replied:
solved

hieule88 commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+                               const DTYPE* __restrict__ output,
+                               const DTYPE* __restrict__ grad, // output_grad
+                               DTYPE* __restrict__ input_grad,
+                               double p,
+                               double n2,
+                               double n2_squared_minus_1,
```
add const

anhskrttt replied:
solved

hieule88 commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+                                         const INPUT_TYPE* __restrict__ output,
+                                         const INPUT_TYPE* __restrict__ grad, // output_grad
+                                         INPUT_TYPE* __restrict__ input_grad,
+                                         double p,
+                                         double n2,
+                                         double n2_squared_minus_1,
```
same

anhskrttt replied:
solved

hieule88 commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+                                          const DTYPE* __restrict__ output,
+                                          const DTYPE* __restrict__ grad, // output_grad
+                                          DTYPE* __restrict__ input_grad,
+                                          double p,
+                                          double n2,
+                                          double n2_squared_minus_1,
+                                          long N,
+                                          long NO,
+                                          long M)
```
same

anhskrttt replied:
solved

hieule88 commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+
+    // pair(row_i, row_j) corresponding to output[k]
+    // e.g. pair(point_A, point_B)
+    long i = n2 - sqrt(n2_squared_minus_1 - 2 * k);
+    long j = k - N * i + i * (i + 1) / 2 + i + 1;
+
+    // Pair of gradients corresponding to pair(i,j)
+    // e.g. pair(input_grad(A,B), input_grad(B,A))
+    // those 2 elements in pair have opposite signs
+    // just similar to dist(A,B) vs. dist(B,A)
+    long ib = j - i - 1;
+    long jb = N - 2 - i;
```
use int64_t or uint64_t (if they cannot be negative). Dont use long

anhskrttt replied:
solved

hieule88 commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+                                          double p,
+                                          double n2,
+                                          double n2_squared_minus_1,
+                                          long N,
+                                          long NO,
+                                          long M)
```
in64_t

anhskrttt replied:
solved

hieule88 commented on src/kernels/MIOpenPdist.cpp:
```diff
@@ -0,0 +1,220 @@
+                        double p,
+                        double n2,
+                        double n2_squared_minus_1,
+                        long N,
+                        long NO,
+                        long M)
+{
```
same

anhskrttt replied:
solved

hieule88 commented on src/kernels/MIOpenSumForward.cpp:
```diff
@@ -0,0 +1,91 @@
+template <typename TI, typename TO, uint32_t IN_NDIMS, uint32_t OUT_NDIMS>
+__device__ void sum_1d_forward(const TI* __restrict__ x,
+                               TO* __restrict__ y,
+                               uint64_t output_numel,
+                               uint64_t reduce_size,
+                               uint64_t inner_size,
+                               uint64_t reduce_dim,
+                               bool nanPropagation,
```
const

anhskrttt replied:
solved

hieule88 commented on src/kernels/MIOpenSumForward.cpp:
```diff
@@ -0,0 +1,91 @@
+
+extern "C" __global__ void Sum1dForward(const INPUT_TYPE* __restrict__ x,
+                                        OUTPUT_TYPE* __restrict__ y,
+                                        uint64_t output_numel,
+                                        uint64_t reduce_size,
+                                        uint64_t inner_size,
+                                        uint64_t reduce_dim,
+                                        bool nanPropagation,
```
same

anhskrttt replied:
solved

anhskrttt commented:
Update 2: I have added `IsImproveOverRocm` to filter non-cont cases that improve the performance. Check final benchmark result in this PR description.

~Updated 1: Having a closer look, I think non-cont is just better in some specific cases, avg_improve for non-cont is only 1.06, compare to avg_improve of cont is 1.47. Check the final benchmark result [here](https://1drv.ms/x/c/f55f630f37e718b8/EXv4O1zOO31IgQmkQyMw5RwBTR-zbeHKMuGgmKzdozibAQ?e=J7GA4Q&nav=MTVfe0IzMTFDQzVGLTlGMjEtNEZFQS1CRjczLTY4MDc2QzVCQjNERH0).~

~I've re-benchmark `PdistBackward` ops with more test cases. it seems to have good performance for both cont and non-cont inputs. Pls see the detail benchmark here.~

anhskrttt commented on driver/dm_pdist.cpp:
```diff
@@ -0,0 +1,40 @@
+
+static Driver* makeDriver(const std::string& base_arg)
+{
+    if(base_arg == "pdist")
+        return new PdistDriver<float, float>();
+    if(base_arg == "pdistfp16")
+        return new PdistDriver<float16, float>();
+    if(base_arg == "pdistbfp16")
+        return new PdistDriver<bfloat16, float>();
```
change `Tref` to `double` because in mloPdistHost you intended to use `double` as the type to calculate on cpu

hieule88 commented on driver/mloPdistHost.hpp:
```diff
@@ -0,0 +1,84 @@
+                    static_cast<float>(input[input_tv.get_tensor_view_idx({j, m})]);
+                float diff = input_first - input_second;
+
+                Tcheck res =
+                    static_cast<Tcheck>(miopen::pdist::backward<float>(diff, grad_k, output_k, p));
+
+                dinputHost[dinput_tv.get_tensor_view_idx({i, m})] += res;
+                dinputHost[dinput_tv.get_tensor_view_idx({j, m})] -= res;
```
dont use Tcheck for calculation
->
```
float res =
                    static_cast<Tcheck>(miopen::pdist::backward<float>(diff, grad_k, output_k, p));
float dinputHost_val_i_m = static_cast<float>(dinputHost[dinput_tv.get_tensor_view_idx({i, m})])
float dinputHost_val_j_m = static_cast<float>(dinputHost[dinput_tv.get_tensor_view_idx({j, m})])

dinputHost[dinput_tv.get_tensor_view_idx({i, m})] = static_cast<Tcheck>(dinputHost_val_i_m  + res)
dinputHost[dinput_tv.get_tensor_view_idx({j, m})] = static_cast<Tcheck>(dinputHost_val_j_m - res)
```

anhskrttt replied:
solved

anhskrttt replied:
can I do this instead, to avoid multiple times of converting?
```
Tcheck res =
                    static_cast<Tcheck>(miopen::pdist::backward<float>(diff, grad_k, output_k, p));

dinputHost[dinput_tv.get_tensor_view_idx({i, m})] =+ res;
dinputHost[dinput_tv.get_tensor_view_idx({j, m})] =- res;
```

Because I saw you done something similar here: https://github.com/moreh-dev/MIOpen/blob/0c707e6e1c5c24e5cc3e27da520e009104a890e8/driver/unsortedsegmentsum_driver.hpp#L65

anhskrttt replied:
i also see this cpu_reducalcuation directly uses `T` for calculation: https://github.com/moreh-dev/MIOpen/blob/b5be72a0c85e035c5b3734c8de91c82c5f65139f/test/gtest/reducecalculation.hpp#L38-L70

Does this mean this code also need fixing?

anhskrttt replied:
UPDATE: ask Hieu-nim again, he said we shouldn't calculate using dtype `Tcheck`. Solved this review

hieule88 commented on test/cpu_pdist.hpp:
```diff
@@ -0,0 +1,79 @@
+                T res = static_cast<T>(miopen::pdist::backward(diff, grad_k, output_k, p));
+
+                ref_input_grad[input_grad_tv.get_tensor_view_idx({i, m})] += res;
+                ref_input_grad[input_grad_tv.get_tensor_view_idx({j, m})] -= res;
```
same

hieule88 commented on src/CMakeLists.txt:
```diff
@@ -167,6 +167,8 @@ set( MIOpen_Source
     op_args.cpp
     operator.cpp
     performance_config.cpp
+    pdist_api.cpp
+    pdist/problem_description.cpp
```
'd' before 'e'

anhskrttt replied:
solved

hieule88 commented on test/gtest/pdist.hpp:
```diff
@@ -0,0 +1,211 @@
+inline std::vector<PdistTestCase> PdistTestConfigs()
+{
+    return {
+
```
rm space

anhskrttt replied:
solved

hieule88 reviewed (commented):
After fix this, you can close the PR. LGTM

anhskrttt commented:
Fully resolve PR reviews.

anhskrttt commented on src/pdist_api.cpp:
```diff
@@ -0,0 +1,135 @@
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include <miopen/common.hpp>
+#include <miopen/miopen.h>
+#include <miopen/errors.hpp>
+#include <miopen/handle.hpp>
+#include <miopen/logger.hpp>
+#include <miopen/tensor_ops.hpp>
+#include <miopen/pdist.hpp>
```
check alphabet. the correct order should be:
```
#include <miopen/common.hpp>
#include <miopen/errors.hpp>
#include <miopen/handle.hpp>
#include <miopen/logger.hpp>
#include <miopen/miopen.h>
#include <miopen/pdist.hpp>
#include <miopen/tensor_ops.hpp>
```

anhskrttt commented on src/include/miopen/pdist/problem_description.hpp:
```diff
@@ -0,0 +1,137 @@
+ *******************************************************************************/
+#pragma once
+
+#include <miopen/activ.hpp>
+#include <miopen/errors.hpp>
+#include <miopen/miopen.h>
```
it seems like those 3 libs are unnecessary. Remove them if not using

anhskrttt commented on src/include/miopen/pdist/invoke_params.hpp:
```diff
@@ -0,0 +1,59 @@
+ *******************************************************************************/
+#pragma once
+
+#include <miopen/common.hpp>
```
unnecessary lib. rm `#include <miopen/common.hpp>`

anhskrttt commented on driver/pdist_driver.hpp:
```diff
@@ -0,0 +1,398 @@
+ *******************************************************************************/
+#pragma once
+
+#include "InputFlags.hpp"
+#include "driver.hpp"
+#include "mloPdistHost.hpp"
+#include "tensor_driver.hpp"
+#include "timer.hpp"
+#include "random.hpp"
+
+#include <../test/tensor_holder.hpp>
+#include <../test/verify.hpp>
+
+#include <miopen/errors.hpp>
+#include <miopen/env.hpp>
+#include <miopen/handle.hpp>
+#include <miopen/miopen.h>
+#include <miopen/tensor.hpp>
+#include <miopen/tensor_view_utils.hpp>
```
correct alphabet order:
```
#include "InputFlags.hpp"
#include "driver.hpp"
#include "mloPdistHost.hpp"
#include "random.hpp"
#include "tensor_driver.hpp"
#include "timer.hpp"

#include <../test/tensor_holder.hpp>
#include <../test/verify.hpp>

#include <miopen/env.hpp>
#include <miopen/errors.hpp>
#include <miopen/handle.hpp>
#include <miopen/miopen.h>
#include <miopen/tensor.hpp>
#include <miopen/tensor_view_utils.hpp>
```

anhskrttt commented on driver/pdist_driver.hpp:
```diff
@@ -0,0 +1,398 @@
+    size_t input_size   = GetTensorSize(inputDesc);
+    size_t output_size  = GetTensorSize(outputDesc);
+    size_t doutput_size = GetTensorSize(doutputDesc);
+    size_t dinput_size  = GetTensorSize(dinputDesc);
```
`dinput_size` is actually the same with `input_size`, might eliminate this if needed
